File: strategy/typescript/glados/agent/keeper/handleball.ts
import * as debug from "base/debug";
import * as Field from "base/field";
import * as Referee from "base/referee";
import { FriendlyRobot } from "base/robot";
import { Position } from "base/vector";
import * as World from "base/world";

import { Behavior, TaskAssignment } from "glados/agent/base/behavior";
import { Objective } from "glados/agent/base/objective";
import { Midfield } from "glados/agent/objective/midfield";
import { MessageType } from "glados/control/messaging";
import * as Physics from "glados/observer/physics";
import { MoveToStaticBall } from "glados/task/attacker/movetostaticball";
import { AggressiveKeeper } from "glados/task/keeper/aggressivekeeper";
import { ChipAway as KeeperChipAway } from "glados/task/keeper/chipaway";
import { Keeper } from "glados/task/keeper/keeper";
import { Pass } from "glados/task/shared/pass";
import * as Attack from "glados/util/attack";
import * as Defense from "glados/util/defense";

export class HandleBall extends Behavior {
	private _pass: {target?: FriendlyRobot, ballPos: Position, time: number} | undefined;
	private _timeBegin: number | undefined = undefined;
	private _objective: Objective | undefined;

	_stop() {
		this._timeBegin = undefined;
		this._objective = undefined;
	}

	check(): Behavior | undefined {
		if (Referee.isStopState() || Referee.isOpponentPenaltyState() || World.GameStage === "PenaltyShootout") {
			return undefined;
		}
		// if a slow ball enters the defense area
		if (Defense.keeperShouldBeMA()) {
			// force being mainAttacker
			this._applyForMainAttacker(undefined, undefined, 2);
		}

		let mainAttackerFlag = this._messaging.receiveTrainer(MessageType.mainAttacker) === this._robot;
		return mainAttackerFlag
			? this
			: undefined;
	}

	_updateTask(): TaskAssignment<typeof Keeper> | TaskAssignment<typeof Pass> | TaskAssignment<typeof KeeperChipAway>
			| TaskAssignment<typeof AggressiveKeeper> | TaskAssignment<typeof MoveToStaticBall> {
		if (!this._objective) {
			this._objective = new Midfield(this._agent);
		}
		this._messaging.sendBroadcast(MessageType.selectedObjective, this._objective);

		let endPos = Physics.ballAtTime(World.Ball, Infinity).pos;
		let startInside = Field.isInFriendlyDefenseArea(World.Ball.pos, -World.Ball.radius - this._robot.radius);
		let endInside = Field.isInFriendlyDefenseArea(endPos, -World.Ball.radius - this._robot.radius);

		// check if there is a danger of a own goal
		let ballDist = Field.distanceToFriendlyGoalLine(World.Ball.pos, 0);
		let robotDist = Field.distanceToFriendlyGoalLine(this._robot.pos, 0);
		let ballBehindKeeper = ballDist < robotDist;

		let suggestions = this._messaging.receive(MessageType.passSuggestion);
		if (startInside && endPos.y < World.Geometry.FriendlyGoal.y + 0.01) {
			// if ball is inside defense area and will enter the goal -> block the ball
			return [Keeper];
		} else if (startInside && endInside && !ballBehindKeeper && suggestions) {
			// if ball is inside defense area and will not leave it -> we have time to act
			// try to find a good pass
			if (this._timeBegin == undefined) {
				this._timeBegin = World.Time;
			}
			const earliestAttackTime = this._messaging.receiveSingleSender(MessageType.earliestAttackTime, true)[1];
			this._pass = Attack.choosePassFromSuggestions(this._robot, suggestions, {
				earliestAttackTime,
				attackPosition: undefined,
				currentPassPos: this._pass?.ballPos,
				considerTiming: false,
				ratePass: Attack.defaultRatePass,
			})[0];
			if (this._pass != undefined) { // check if there is a good pass, else chip away
				if (this._pass.target) {
					if (this._task != undefined && this._task instanceof Pass) {
						this._task.updateTarget(this._pass.target, this._pass.ballPos, true);
					}
					this._messaging.sendBroadcast(MessageType.passInfo, [{ target: this._pass.target,
						ballPos: this._pass.ballPos, time: this._pass.time}]);
				}

				return [
					Pass, [{
						targetRobot: this._pass.target,
						targetPos: this._pass.ballPos,
						chip: true,
						highPrecision: true,
					}],
				];
			}
			if (World.Time - this._timeBegin < 7) {
				return [MoveToStaticBall];
			} else {
				return [KeeperChipAway];
			}
		} else {
			// if inside and ball will leave or outside -> get rid of the ball
			return [AggressiveKeeper];
		}
	}
}

File: strategy/lua/marvin/agent/keeper/handleball.lua
--[[***********************************************************************
*   Copyright 2026 Robotics Erlangen e.V.                                 *
*   http://www.robotics-erlangen.de/                                      *
*   info@robotics-erlangen.de                                             *
*                                                                         *
*   This program is free software: you can redistribute it and/or modify  *
*   it under the terms of the GNU General Public License as published by  *
*   the Free Software Foundation, either version 3 of the License, or     *
*   any later version.                                                    *
*                                                                         *
*   This program is distributed in the hope that it will be useful,       *
*   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
*   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
*   GNU General Public License for more details.                          *
*                                                                         *
*   You should have received a copy of the GNU General Public License     *
*   along with this program.  If not, see <http://www.gnu.org/licenses/>. *
*************************************************************************]]

local Base = require "agent/base/behavior"
local HandleBall = Class("Agent.Keeper.HandleBall", Base)

local Field = require "../base/field"
local Referee = require "../base/referee"
local World = require "../base/world"

local Ball = require "observer/ball"
local Physics = require "observer/physics"
local AggressiveKeeper = require "task/keeper/aggressivekeeper"
local Keeper = require "task/keeper/keeper"
local KeeperChipAway = require "task/keeper/chipaway"
local Pass = require "task/shared/pass"
local Attack = require "util/attack"

function HandleBall:_init()
	self._pass = nil
	self._hysteresis = false
end

function HandleBall:behindCenterbacks(object)
	local hyst = self._hysteresis and 0.1 or 0
	local defenseDistance = self._robot.radius + self._robot.shootRadius + hyst
	return Field.distanceToFriendlyDefenseArea(object.pos, object.radius) < defenseDistance
end

function HandleBall:check()
	if Referee.isStopState() or Referee.isOpponentPenaltyState() or World.GameStage == "PenaltyShootout" then
		return false
	end
	-- if a slow ball enters the defense area
	local active = self:behindCenterbacks(World.Ball) and Ball.isSlowBall()
	if active then
		-- force being mainAttacker
		self._hysteresis = true
		self:_applyForMainAttacker(nil, nil, 2)
	else
		self._hysteresis = false
	end

	local mainAttackerFlag = self._inbox.mainAttacker().trainer == self._robot
	return mainAttackerFlag
end

function HandleBall:_updateTask()
	local endPos = Physics.ballAtTime(World.Ball, math.huge).pos
	local startInside = Field.isInFriendlyDefenseArea(World.Ball.pos, -World.Ball.radius-self._robot.radius)
	local endInside = Field.isInFriendlyDefenseArea(endPos, -World.Ball.radius-self._robot.radius)

	-- check if there is a danger of a own goal
	local ballDist = Field.distanceToFriendlyGoalLine(World.Ball.pos, 0)
	local robotDist = Field.distanceToFriendlyGoalLine(self._robot.pos, 0)
	local ballBehindKeeper = ballDist < robotDist

	if startInside and endPos.y < World.Geometry.FriendlyGoal.y + 0.01 then
		-- if ball is inside defense area and will enter the goal -> block the ball
		return Keeper
	elseif startInside and endInside and not ballBehindKeeper and self._inbox.passSuggestion() then
		-- if ball is inside defense area and will not leave it -> we have time to act
		-- try to find a good pass
		self._pass = Attack.choosePassFromSuggestions(self._robot, self._inbox.passSuggestion(), self._pass and self._pass.ballPos, false)
		if self._pass then --check if there is a good pass, else chip away
			return Pass, { self._pass.target, self._pass.ballPos, true }
		else
			return KeeperChipAway
		end
	else
		-- if inside and ball will leave or outside -> get rid of the ball
		return AggressiveKeeper
	end
end

return HandleBall
